crates/matrix-sdk/src/sliding_sync/mod.rs

```diff
@@ -483,15 +486,24 @@ impl SlidingSync {
             sticky_params.data().extensions.to_device.enabled == Some(true)
         };
 
-        // Set the to_device token if the extension is enabled.
-        if to_device_enabled {
+        // Set the to_device token if the extension is enabled, and/or override the `pos`
+        // field from the database.
+        if to_device_enabled || self.inner.restore_pos_from_database {
             let lists = self.inner.lists.read().await;
 
             if let Some(restored_fields) =
                 restore_sliding_sync_state(&self.inner.client, &self.inner.storage_key, &lists)
                     .await?
             {
-                request.extensions.to_device.since = restored_fields.to_device_token;
+                if to_device_enabled {
+                    request.extensions.to_device.since = restored_fields.to_device_token;
+                }
+                if self.inner.restore_pos_from_database {
+                    if request.pos != restored_fields.pos {
```

Comment:
Did you do that so that every time we try to build a request, we read the value from the storage? So if someone else writes into the cache, we will use the latest version from the cache? Do you also bet that every time we get a response from the server, we save `SlidingSync` in the cache, so… we will just land back on tracks with the previous workflow:

* response received from the server
* save `SlidingSync` into the cache
* next request uses `pos` from the response
* send the request

vs.


* response received from the server
* save `SlidingSync` into the cache
* next request uses `pos` from the response
* if `pos` is different from what we have in the storage, let's use it in the request
* send the request

Is it the spirit?

Reply:
I think so, I'm not so sure what you're asking here, sorry.

The new flow is:

- a response is received from the server
- we update the in-memory pos based on that response,
- *then* cache the pos into disk storage

When generating the next request:

- if configured to do so, read the pos from the storage
  - either it was us who write it, and then it'll be the same as the one we have in memory
  - or another process ran a sync in the background, and updated the pos in db, so we must use the value read from the db

In both cases, getting the value from the DB is the right thing to do.